miner/src/pool/queue.rs

```diff
@@ -0,0 +1,440 @@
+
+		self.pool.read().pending_from_sender(state_readiness, address)
+			.last()
+			.map(|tx| tx.signed().nonce + 1.into())
+	}
+
+	/// Retrieve a transaction from the pool.
+	///
+	/// Given transaction hash looks up that transaction in the pool
+	/// and returns a shared pointer to it or `None` if it's not present.
+	pub fn find(
+		&self,
+		hash: &H256,
+	) -> Option<Arc<pool::VerifiedTransaction>> {
+		self.pool.read().find(hash)
+	}
+
+	/// Remove a set of transactions from the pool.
+	///
+	/// Given an iterator of transaction hashes
+	/// removes them from the pool.
+	/// That method should be used if invalid transactions are detected
+	/// or you want to cancel a transaction.
+	pub fn remove<'a, T: IntoIterator<Item = &'a H256>>(
+		&self,
+		hashes: T,
+		is_invalid: bool,
+	) -> Vec<Option<Arc<pool::VerifiedTransaction>>> {
+		let results = {
+			let mut pool = self.pool.write();
+
+			hashes
+				.into_iter()
+				.map(|hash| pool.remove(hash, is_invalid))
+				.collect::<Vec<_>>()
+		};
+
+		if results.iter().any(Option::is_some) {
+			self.cached_pending.write().clear();
+		}
+
+		results
+	}
+
+	/// Clear the entire pool.
+	pub fn clear(&self) {
+		self.pool.write().clear();
+	}
+
+	/// Penalize given senders.
+	pub fn penalize<'a, T: IntoIterator<Item = &'a Address>>(&self, senders: T) {
+		let mut pool = self.pool.write();
+		for sender in senders {
+			pool.update_scores(sender, ());
+		}
+	}
+
+	/// Returns gas price of currently the worst transaction in the pool.
+	pub fn current_worst_gas_price(&self) -> U256 {
+		match self.pool.read().worst_transaction() {
+			Some(tx) => tx.signed().gas_price,
+			None => self.options.read().minimal_gas_price,
+		}
+	}
+
+	/// Returns a status of the queue.
+	pub fn status(&self) -> Status {
+		let pool = self.pool.read();
+		let status = pool.light_status();
+		let limits = pool.options();
+		let options = self.options.read().clone();
+
+		Status {
+			options,
+			status,
+			limits,
+		}
+	}
+
+	/// Check if there are any local transactions in the pool.
+	///
+	/// Returns `true` if there are any transactions in the pool
+	/// that has been marked as local.
+	///
+	/// Local transactions are the ones from accounts managed by this node
+	/// and transactions submitted via local RPC (`eth_sendRawTransaction`)
+	pub fn has_local_pending_transactions(&self) -> bool {
+		self.pool.read().listener().0.has_pending()
+	}
+
+	/// Returns status of recently seen local transactions.
+	pub fn local_transactions(&self) -> BTreeMap<H256, pool::local_transactions::Status> {
+		self.pool.read().listener().0.all_transactions().iter().map(|(a, b)| (*a, b.clone())).collect()
+	}
+
+	/// Add a callback to be notified about all transactions entering the pool.
+	pub fn add_listener(&self, f: Box<Fn(&[H256]) + Send + Sync>) {
+		let mut pool = self.pool.write();
+		(pool.listener_mut().1).0.add(f);
+	}
+}
+
+
+fn convert_error(err: txpool::Error) -> transaction::Error {
```

Comment:
why did you implement `fn convert_error` instead of `into`?

Reply:
neither `txpool::Error` nor `transaction::Error` is defined in that crate.